lib/mlton/basic/stream-parser.fun

```diff
@@ -0,0 +1,257 @@
+(* Copyright (C) 2017 Jason Carr.
+ *
+ * MLton is released under a BSD-style license.
+ * See the file MLton-LICENSE for details.
+ *)
+
+functor StreamParser(S: STREAM_PARSER_STRUCTS):STREAM_PARSER = 
+struct
+
+infix 1 <|> >>=
+infix 2 <&>
+infix  3 <*> <* *> 
+infixr 4 <$> <$$> <$$$> <$
+
+
+
+open S
+
+type location = {line: int, column: int}
+type info = string
+(* this is our state representation for readers, but info is unchanging *)
+datatype 'b result = Success of 'b * (char * location) Stream.t
+                   | Failure of string list (* expected values *)
+                   | FailCut of string list (* as failure, but the
+                   closest upstream choice point won't try other options, and
+                   their errors will be silenced *)
+type state = info * (char * location) Stream.t
+type 'b t = 
+   state -> 'b result
+
+
+fun indexStream({line, column}, s) = 
+   case Stream.force s of
+      NONE => Stream.empty ()
+    | SOME(h, r) =>
+         Stream.cons((h, {line=line, column=column}), 
+            Stream.delay(fn () =>
+               if h = #"\n"
+               then 
+                  indexStream({line=line+1, column=0}, r)
+               else
+                  indexStream({line=line, column=column+1}, r)
+            )
+         )
+
+fun doFail([]) = raise Fail("Parse error")
+  | doFail([msg]) = raise Fail ("Parse error: Expected " ^ msg)
+  | doFail(msgs) = raise Fail ("Parse error: Expected one of \n" ^
+       (String.concat(List.map(msgs, fn x => x ^ "\n\n"))))
+
+fun doParse(p : 'b t, info, s) = 
+   case p (info, indexStream({line=1, column=1}, s)) 
+   of Success (b, _) => b
```

Comment:
I think the current functionality of partial results is the best here. EOF can be checked through failing(next) (or we could add a primitive), and this behavior is consistent with other parser combinator libraries. Although no function right now uses partial results, this could implement it without stretching the interface too much.